File: genetic-packer/src/main/java/genetic/packer/mapper/PackerContextMapper.java
package genetic.packer.mapper;

import genetic.packer.Packer;
import genetic.packer.evolution.generation.GeneticContext;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

import java.util.function.Function;

/**
 * @author piotr.larysz
 */
@Mapper
public abstract class PackerContextMapper implements Function<Packer.Context, GeneticContext> {

    @Override
    @Mapping(target = "chromosomeSize", expression = "java(context.getEmbryo().getBoxes().size())")
    @Mapping(target = "numberOfEliteIndividuals", defaultValue = "1")
    public abstract GeneticContext apply(Packer.Context context);
}

File: genetic-packer/src/main/java/genetic/packer/mapper/PackerContextMapper.kt
package genetic.packer.mapper

import genetic.Evolution
import genetic.api.generation.GenerationContext
import genetic.packer.evolution.generation.dto.Embryo

class PackerContextMapper : (Evolution.Context<Embryo>) -> GenerationContext<Embryo> {

  override fun invoke(context: Evolution.Context<Embryo>) = GenerationContext(
      embryo = context.embryo,
      chromosomeSize = context.embryo.boxes.count(),
      numberOfEliteIndividuals = context.numberOfEliteIndividuals,
      generationSize = context.generationSize
  )
}